File: src/main/kotlin/day5/Day5.kt
package day5

import inputLines

fun getPos(input: String): Pair<Int, Int> =
    input.take(7).replace("F", "0").replace("B", "1").toInt(2) to
    input.drop(7).replace("L", "0").replace("R", "1").toInt(2)


fun main() {
    val r: List<Int> = inputLines(day = 5)
            .map(::getPos)
            .map { (row, col) -> row*8 + col }
            .sorted()
    println(r.zip(r.drop(1)).firstOrNull { (a, b) -> a + 1 != b }!!.first)
}

File: src/main/kotlin/day5/day5.hs
import Data.List



main = readFile "inputs/2020/5.txt" >>= print . main2


getID :: String -> Int
getID str = row * 8 + col
    where row = fromBinary [if c == 'F' then '0' else '1' | c <- take 7 str]
          col = fromBinary [if c == 'L' then '0' else '1' | c <- drop 7 str]

fromBinary :: [Char] -> Int
fromBinary [] = 0
fromBinary ['1'] = 1
fromBinary ['0'] = 0
fromBinary xs = (fromBinary $ init xs) * 2 + case last xs of { '0' -> 0; '1' -> 1; }

main1 :: String -> Int
main1 x = maximum $ getID <$> lines x

main2 :: String -> Int
main2 x = head [ a | (a, b) <- zip l (tail l), a+ 1 /= b]
    where l = sort $ getID <$> lines x